### PR TITLE
Clean render return type

### LINKED ISSUE #4063: Tidy the return type of `App.render`
Currently the return type of `App.render` is `RenderableType`; in our docs and in most places we talk about `App.render` returning a `RenderResult`. `RenderResult` is an alias for `RenderableType` and is a top-level value in `textual.app`.

This can cause a lint/type warning in dev's apps if they let their IDE complete the signature when writing their own `render` method, as it will often complete to something like this:

```python
    def render(self) -> RenderableType:
        return super().render()
```

which in turn will auto-import `RenderableType` from `textual.app`. This can then result in a warning like this:

Screenshot 2024-01-23 at 14 16 46

We should change `App.render` so that it's typed as returning `RenderResult` rather than `RenderableType`, then such completion will use the type that *is* exported from `textual.app`.

Also note that [the code in the compose/render HOWTO seems to have fallen foul of this](https://textual.textualize.io/how-to/render-and-compose/) and has that type warning.

### PR DESCRIPTION
**Pull request by davep**
This PR tweaks the type hinting for the `render` methods of various widgets, to ensure they follow the same hinting as is used [in our documentation](https://textual.textualize.io/guide/widgets/#custom-widgets); the aim being to make the code a little less surprising to the reader and also to ensure that any type hint information shown in an IDE, or auto-completed when inheriting from a Textual widget and overriding the `render` method, matches what would be expected.

Fixes #4063.

### CODE REVIEW

File: src/textual/widget.py [hunk 1]
```diff
@@ -32,15 +32,18 @@
     ConsoleRenderable,
     JustifyMethod,
     RenderableType,
-    RenderResult,
-    RichCast,
 )
+from rich.console import RenderResult as RichRenderResult
```
Comment (rodrigogiraoserrao) [inline src/textual/widget.py:36]: Why are these two imports separated from the other `rich.console` imports?
Why not the suggestion I show below?

```py
from rich.console import (
    Console,
    ConsoleOptions,
    ConsoleRenderable,
    JustifyMethod,
    RenderableType,
    RenderResult as RichRenderResult,
    RichCast,
)
```

File: src/textual/widgets/_input.py [hunk 2]
```diff
@@ -47,7 +52,7 @@ def __init__(self, input: Input, cursor_visible: bool) -> None:
 
     def __rich_console__(
         self, console: "Console", options: "ConsoleOptions"
-    ) -> "RenderResult":
+    ) -> "RichRenderResult":
```
Comment (rodrigogiraoserrao) [inline src/textual/widgets/_input.py:55]:
```suggestion
    ) -> RichRenderResult:
```

**Review by rodrigogiraoserrao — APPROVED** [review]
Looking at this “simple” PR, I can't help but feel like you spent a _lot_ of time making sure you were getting all of the instances and to make sure you don't mix the app and Rich's `RenderResult` 😬

File: src/textual/widget.py (hunk 1 above)
Comment (davep) [reply src/textual/widget.py:36]: I'll let you have that argument with black and/or whatever else it is that rewrote my code from that.

**Comment by davep** [comment]
> Looking at this “simple” PR, I can't help but feel like you spent a lot of time making sure you were getting all of the instances and to make sure you don't mix the app and Rich's RenderResult 😬

Yup, that was fun. As I worked through this too I also got to thinking that, in retrospect, it might have been nicer if `RenderResult` and `ComposeResult` originated in `textual.types` too. Possibly. Maybe. Perhaps.

File: src/textual/widget.py (hunk 1 above)
Comment (rodrigogiraoserrao) [reply src/textual/widget.py:36]: Ah ok, I understand 🤣

**Pull request merged**